### PR TITLE
introduce `_app.mdx` for better performance and smallest `.next/static/chunks` size

### PR DESCRIPTION
**Pull request by B2o5T**
I am pretty sure, this will fix https://github.com/shuding/nextra/issues/1131

also, it opens the opportunity to use react context in future refactor of v3 instead of `useInternals`?

## Update

> I guess there is some race condition for `_app.mdx` sometimes `next build` passe and sometimes not

image

✅ Fixed here https://github.com/shuding/nextra/pull/1463

## before
```
 dmytro@MacBook-Pro  ~/Desktop/nextra   main  find examples/blog/.next/static/chunks -ls | awk '{sum += $7} END {print sum}'
435979
 dmytro@MacBook-Pro  ~/Desktop/nextra   main  find examples/docs/.next/static/chunks -ls | awk '{sum += $7} END {print sum}'
980947
 dmytro@MacBook-Pro  ~/Desktop/nextra   main  find examples/swr-site/.next/static/chunks -ls | awk '{sum += $7} END {print sum}'
4060262
 dmytro@MacBook-Pro  ~/Desktop/nextra   main  find docs/.next/static/chunks -ls | awk '{sum += $7} END {print sum}' 
1342697
```

## after
```
 dmytro@MacBook-Pro  ~/Desktop/nextra   _app  find examples/blog/.next/static/chunks -ls | awk '{sum += $7} END {print sum}'    
429701
 dmytro@MacBook-Pro  ~/Desktop/nextra   _app  find examples/docs/.next/static/chunks -ls | awk '{sum += $7} END {print sum}'
958154
 dmytro@MacBook-Pro  ~/Desktop/nextra   _app  find examples/swr-site/.next/static/chunks -ls | awk '{sum += $7} END {print sum}'
3247164
 dmytro@MacBook-Pro  ~/Desktop/nextra   _app  find docs/.next/static/chunks -ls | awk '{sum += $7} END {print sum}'     
1254104
```

## diff

| example-blog | example-docs | swr-site | nextra website |
|--------------|--------------|----------|----------------|
| 1.45%        | 2.35%        | 20.25%   | 6.6%          |

### CODE REVIEW

**Review by promer94 — COMMENTED** [review]
I like the idea we only need to inject `pageMap` once  ! 
I am also doing some exterimental about this optimization.

## Concerns
_app.js is a important entry for a next.js app. People used to do a lot of customizations around this file. 

* Analytics
```tsx
import type { AppProps } from 'next/app';
import { Analytics } from '@vercel/analytics/react';

function MyApp({ Component, pageProps }: AppProps) {
  return (
    <>
      <Component {...pageProps} />
      <Analytics />
    </>
  );
}

export default MyApp;
```

* Auth

```tsx
import { SessionProvider } from "next-auth/react"
export default function App({
  Component,
  pageProps: { session, ...pageProps },
}) {
  return (
    <SessionProvider session={session}>
      <Component {...pageProps} />
    </SessionProvider>
  )
}
```

It seems hard to do the same customization in app.mdx ?


## Possible Alternatives
Maybe we could build
* a webpack plugin used to create a virtual file which contains the pageMaps and other meta data
* a special loader just for _app.js so it could import the virtual file

**Comment by B2o5T** [comment]
> _app.js is a important entry for next.js. Developers used to do a lot of customizations around this file.

totally agree

the first concern can be already done

https://github.com/shuding/nextra/blob/785e67c821cb51f8e180165bfc0d7ce9884be749/packages/nextra/src/setup-underscore-app.tsx#L21-L24

```mdx
// _app.mdx
import { Analytics } from '@vercel/analytics/react'

<Analytics />
```

for the second one need to think about hove to have possibility to wrap everything with user provider

File: packages/nextra/src/loader.ts [hunk 1]
```diff
@@ -243,43 +257,38 @@ async function loader(
...
-          descriptor
-        )}]`
-    )
-    .join(',')}] : []
-})`
+  pageOptsChecksum: ${pageOptsChecksum},
+  dynamicMetaModules: typeof window === 'undefined' ? [${dynamicMetaModules}] : []
+})
+
+export { default } from 'nextra/layout'`
```
Comment (B2o5T) [inline packages/nextra/src/loader.ts:291]: I think `nextra/layout` we can move directly to `nextra/setup-page` to have less file size

File: packages/nextra/src/loader.ts [hunk 2]
```diff
@@ -243,43 +257,38 @@ async function loader(
...
+        `[import(${JSON.stringify(descriptor.metaFilePath)}), ${JSON.stringify(
+          descriptor
+        )}]`
+    )
+    .join(',')
 
   return `import { setupNextraPage } from 'nextra/setup-page'
 import __nextra_layout from '${layout}'
 ${themeConfigImport}
 ${katexCssImport}
```
Comment (B2o5T) [inline packages/nextra/src/loader.ts:276]: these imports we can inject also in `_app.mdx` to yet reduce bundle size, but in another PR ;)

**Comment by promer94** [comment]
> ```tsx
> // _app.mdx
> import { SessionProvider } from "next-auth/react"
> export default function App({
>   Component,
>   pageProps: { session, ...pageProps },
> }) {
>   return (
>     <SessionProvider session={session}>
>       <Component {...pageProps} />
>     </SessionProvider>
>   )
> }
> ```

This is ok but not ideal. Since user will lost almost all the code intellisense and autocomplete in mdx.  

I know `mdx.experimentalLanguageServer:` could help to improve this but it is still in its early stage.

**Comment by B2o5T** [comment]
> > ```tsx
> > // _app.mdx
> > import { SessionProvider } from "next-auth/react"
> > export default function App({
> >   Component,
> >   pageProps: { session, ...pageProps },
> > }) {
> >   return (
> >     <SessionProvider session={session}>
> >       <Component {...pageProps} />
> >     </SessionProvider>
> >   )
> > }
> > ```
> 
> This is ok but not ideal. Since user will lost almost all the code intellisense and autocomplete in mdx.
> 
> I know `mdx.experimentalLanguageServer:` could help to improve this but it is still in its early stage.

you can also write your code in typescript and import in `_app.mdx`

```jsx
// _app.mdx
export { default } from '../../myUnderscoreAppWrittenInTs'
```

File: packages/nextra/src/loader.ts [hunk 3]
```diff
@@ -235,6 +233,22 @@ async function loader(
     // necessary to include them in the bundle.
     pageOpts = transformPageOpts(pageOpts)
   }
+  if (HAS_UNDERSCORE_APP_MDX_FILE) {
```
[the remark this answers is not in the document]
Comment (B2o5T) [reply packages/nextra/src/loader.ts:236]: this is a nice improvement, I was thinking about it also, let's do it but in another PR

**Comment by danez** [comment]
In all fairness, this should have been a breaking change. It made my website not build for some time and  took me a while to figure out that I had to run a build locally to create`_app.mdx` and commit it.

**Comment by B2o5T** [comment]
@danez Can you please open a separate issue for this? It's hard to properly track this in the comments of an already merged PR.

**Pull request merged**